### ISSUE TITLE
Uncaught Error: JIT compiler unavailable

### ISSUE THREAD

**Issue by stevethemacguy** [angular/angular-cli#21704]
# 🐞 Bug report

### Command (mark with an `x`)

N/A. The app is built in AOT mode using `@ngtools/webpack@12.2.4` and a custom Webpack config file.

### Is this a regression?
Yes, the previous version in which this bug was not present was: `@ngtools/webpack@11.2.14`

### Description

The AOT build in Webpack 5 (and also Webpack 4) succeeds with no errors, but the app fails to render and the browser shows the following error in the console:

```
Uncaught Error: JIT compiler unavailable
    at qn (app.d58833d2c8eb48cf8f29.js:2)
```

Releated app.js source code:
```
function qn(e) {
  const t = Q.ng;
  if (t && t.\u0275compilerFacade) return t.\u0275compilerFacade;
  throw new Error('JIT compiler unavailable')
}
```

## 🔬 Minimal Reproduction

I created a new Angular CLI project using a custom webpack config to try to mimic our environment, including identical tsconfig.json and main.ts files, but I was unable to reproduce the error. The only change we made was upgrading all Angular packages from `11.2.14` to `12.2.4`. No changes were made to our webpack config files or any other code.

I thought this might be a conflict with Angular 12 and Webpack 4 (i.e. `@ngtools/webpack@11.2.14`), so I upgraded to Webpack 5 (i.e. upgraded @ngtools/webpack, zone.js, and all other packages to the "latest" versions). Both the Webpack 4 and Webpack 5 builds succeed with no errors, but the Angular run-time error still occurs.

Note: The error occurs when running the AOT build locally on an Express server and when deploying it to our Nginx Server. Our Webpack config does _not_  include webpack-dev-server.

Here is the @ngtools/webpack config:

```
new AngularWebpackPlugin({
      tsconfig: helpers.root('src', 'prod.tsconfig.json'),
      jitMode: false
}),
```

## 🔥 Exception or Error

<pre><code>
Uncaught Error: JIT compiler unavailable
    at qn (app.d58833d2c8eb48cf8f29.js:2)
</code></pre>

## 🌍 Your Environment

<pre><code>
Angular CLI: 12.2.4
Node: 14.16.1
Package Manager: npm 6.14.12
OS: darwin x64

Angular: 12.2.4
... animations, cdk, cli, common, compiler, compiler-cli, core
... forms, localize, material, platform-browser
... platform-browser-dynamic, router

Package                           Version
-----------------------------------------------------------
@angular-devkit/architect         0.1202.4
@angular-devkit/build-optimizer   0.1202.4
@angular-devkit/core              12.2.4
@angular-devkit/schematics        12.2.4
@ngtools/webpack                  12.2.4
@schematics/angular               12.2.4
rxjs                              6.6.7
typescript                        4.3.2
webpack                           5.51.1
</code></pre>

**Anything else relevant?**
Someone else posted a similar error on [Stack Overflow](https://stackoverflow.com/questions/68942317/jit-compiler-unavailable-after-upgrading-angular-from-11-to-12).

**Comment by alan-agius4** [comment]
Unfortunately without a reproduction it's hard to tell what is happening.

**Comment by stevethemacguy** [comment]
@alan-agius4 The real problem turned out to be the `ng-click-outside` plugin, which is no longer compatible with the Angular 12 AOT compiler. Unfortunately, a more helpful error was never shown because of this line in the core compiler:

```
if (typeof ngDevMode === 'undefined' || ngDevMode) {
   ....
}
else {
    throw new Error('JIT compiler unavailable');
}

// See @angular/core/__ivy_ngcc__/fesm2015/core.js
```

When the [Terser Webpack Plugin](https://webpack.js.org/plugins/terser-webpack-plugin/) config includes `ngDevMode: false`, the compiler throws the generic 'JIT compiler unavailable' error instead of a useful error. In my case, once I removed the TerserPlugin flag, the _real_  compiler error was displayed:

![JIT Compiler Error](https://user-images.githubusercontent.com/1767135/132426519-646dd567-9fc7-4505-bbbf-c13eacdf4651.png)

Setting `ngDevMode: false` is recommended by the [Angular 12 Ivy Compatibility Guide](https://angular.io/guide/ivy-compatibility#payload-size-debugging), so other developers might encounter this issue. Is there a good reason to throw the generic error when 'ngDevMode' is set? I think it would be helpful to add a warning to the [Update Guide](https://update.angular.io/) about using this flag with Angular 12. It would have saved me a lot of time.

Also, I fixed the issue by adding `import '@angular/compiler';` to main.ts, but that doesn't seem like a good idea in production. What are the implications of including the compiler? Does it change the output bundle size?

**Comment by alan-agius4** [comment]
> I fixed the issue by adding import '@angular/compiler'; to main.ts, but that doesn't seem like a good idea in production. What are the implications of including the compiler? Does it change the output bundle size?

Yes, this does increase the bundle size quite a lot. `ng-click-outside` also seems to be unmaintained, and therefore one should seek to use an alternate package. 

> When the Terser Webpack Plugin config includes ngDevMode: false, the compiler throws the generic 'JIT compiler unavailable' error instead of a useful error. In my case, once I removed the TerserPlugin flag, the real compiler error was displayed:

This error should be shown also during development when using AOT without terser, isn't this the case?

> Is there a good reason to throw the generic error when 'ngDevMode' is set? 
Yeah, this is because JIT and development context and information is removed for production builds to reduce the bundle size even further. 

> I think it would be helpful to add a warning to the Update Guide about using this flag with Angular 12. It would have saved me a lot of time.

This seems to be quite a rare care, and therefore it don't think it warrants to be included in the update guide.

**Comment by stevethemacguy** [comment]
Thank you for the prompt response. Here are the two errors for reference:

**Generic**: "JIT compiler unavailable"
**Useful**:   "JIT compilation failed...xxx needs to be compiled using the JIT compiler, but '@angular/compiler' is not available."

> This error should be shown also during development when using AOT without terser, isn't this the case?

The **Useful** error (which mentions the broken plugin) is shown in these cases:
1 ) `optimization.minimize: false`
2 ) `optimization.minimize: true` + TerserPlugin + `ngDevMode: true`
3 ) `optimization.minimize: true` + TerserPlugin + `ngDevMode` not specified

The **Generic** error is shown when
1 ) `optimization.minimize: true` + TerserPlugin + `ngDevMode: false`
> Is there a good reason to throw the generic error when `ngDevMode` is set?

From a user standpoint, it was surprising to see "JIT compiler unavailable" when using AOT because I assumed the JIT compiler wasn't used at all in AOT mode. More importantly, regardless of whether the JIT compiler failed or was 'unavailable', the  **Useful** error was more helpful in both cases since it told me how to fix the bug (i.e. by importing @angular/compiler). 

Since the **Generic** error is only thrown when `ngDevMode` is false, which we know was explicitly set by the user, perhaps it should say something like "The JIT compiler is unavailable. If you meant to include it, either import '@angular/compiler' or set ngDevMode to true." As it's currently written, the **Generic** error didn't help either of us know how to fix the problem.

> This seems to be quite a rare care, and therefore it don't think it warrants to be included in the update guide.

Using the Ivy Compatibility Guide with TerserPlugin and several plugins doesn't seem like a rare case, but re-wording the **Generic** error should take care of the issue without updating anything else.

**Comment by alan-agius4** [comment]
In general you don't need and shouldn't import `@angular/compiler` directly. Typically you need to do so in extreme cases when your application is using both JIT and AOT (IE: hybrid version). Unless you are doing so, this indicates that there might be a problem with either your application or a library.

> Using the Ivy Compatibility Guide with TerserPlugin and several plugins doesn't seem like a rare case, but re-wording the Generic error should take care of the issue without updating anything else.

Note, the Angular tooling team doesn't support custom Webpack configurations. By choosing to extend the Webpack configuration you effectively moved away from the happy path.

Overall. I do think that the error could be improved and restore the generic error message with the one prior to https://github.com/angular/angular/pull/42693

//cc @JoostK
